fasttld/FastTLDExtract.py

```diff
@@ -49,84 +88,83 @@ def _trie_construct(self, exclude_private_suffix, file_path=''):
-                        '_END': True
-                    }
-                })
+                tld_trie[suffix] = {'_END': True}
         for key, val in tld_trie.items():
-            if len(tld_trie[key]) == 1:
-                if '_END' in val:
-                    tld_trie.update({
-                        key: True
-                    })
+            if len(val) == 1 and '_END' in val:
+                tld_trie[key] = True
         return tld_trie
-    
+
     def __call__(self, *args, **kwargs):
         return self.extract(*args, **kwargs)
 
-    def extract(self, input, subdomain=True, format=False):
+    def extract(self, raw_url, subdomain=True, format=False):
         """
         Extract suffix and subdomain from a Domain.
-        :param input: 
-        :param subdomain: Output options. This option will reduce efficiency. Maybe 10% 
-        :param format: To format input string.
+        :param raw_url:
+        :param subdomain: Output options. This option will reduce efficiency. Maybe 10%
+        :param format: To format raw_url string.
         :return: Tuple(subdomain, domain, suffix, domain_name)
         >>> FastTLDExtract.extract('www.google.com.hk', subdomain=True)
         >>> ('www', 'google', 'com.hk', 'google.com.hk')
+
+        >>> FastTLDExtract.extract('127.0.0.1', subdomain=True)
+        >>> ('', '127.0.0.1', '', '127.0.0.1')
         """
-        ret_subdomain = ''
-        ret_domain = ''
-        ret_suffix = ''
-        ret_domain_name = ''
+        ret_subdomain = ret_domain = ret_suffix = ret_domain_name = ''
         if format:
-            input = self.format(input)
-        labels = input.split('.')
+            raw_url = self.format(raw_url)
+
+        # Borrowed from tldextract library (https://github.com/john-kurkowski/tldextract)
+        # Use regex to strip raw_url of scheme subcomponent and anything after host subcomponent
+        # Reference: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax
+        netloc = (
+            SCHEME_RE.sub("", raw_url)
+            .partition("/")[0]
+            .partition("?")[0]
+            .partition("#")[0]
+            .split("@")[-1]
+            .partition(":")[0]
+            .strip()
+            .rstrip(".")
+        )
+        # Determine if raw_url is an IP address
+        if len(netloc) != 0 and looks_like_ip(netloc):
+            return ("", netloc, "", netloc)
+        labels = netloc.split(".")
         labels.reverse()
         node = self.trie  # define the root node
         suffix = []
-
-        for index, label in enumerate(labels):
+        for label in labels:
             if node is True:
                 # This node is an end node.
                 ret_domain = label
                 break
 
             # This node has sub-nodes and maybe an end-node.
-            # eg. us -> (us, gov.us)
+            # eg. cn -> (cn, gov.cn)
             if '_END' in node:
-                if index < len(labels):
```

Comment:
`if index < len(labels)` is redundant here as it is already guaranteed to be `True`.